internal/controllers/inventory_controller.go

```diff
@@ -371,10 +374,156 @@ func (t *reconcilerTask) setupAlarmSubscriptionServerConfig(ctx context.Context,
+			return nil, fmt.Errorf("failed to get configmap: %s", err.Error())
+		}
+
+		caBundle, err = utils.GetConfigMapField(cm, "ca-bundle.pem")
+		if err != nil {
+			return nil, fmt.Errorf("failed to get certificate bundle from configmap: %s", err.Error())
+		}
+	}
+
+	config := utils.OAuthClientConfig{
+		CaBundle: []byte(caBundle),
+	}
+
+	oAuthConfig := t.object.Spec.SmoConfig.OAuthConfig
+	if oAuthConfig != nil {
+		clientSecrets, err := utils.GetSecret(ctx, t.client, oAuthConfig.ClientSecretName, t.object.Namespace)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get client secret: %w", err)
+		}
+
+		clientId, err := utils.GetSecretField(clientSecrets, "client-id")
+		if err != nil {
+			return nil, fmt.Errorf("failed to get client-id from secret: %s, %w", oAuthConfig.ClientSecretName, err)
+		}
+
+		clientSecret, err := utils.GetSecretField(clientSecrets, "client-secret")
+		if err != nil {
+			return nil, fmt.Errorf("failed to get client-secret from secret: %s, %w", oAuthConfig.ClientSecretName, err)
+		}
+
+		config.ClientId = clientId
+		config.ClientSecret = clientSecret
+		config.TokenUrl = fmt.Sprintf("%s%s", oAuthConfig.Url, oAuthConfig.TokenEndpoint)
+		config.Scopes = oAuthConfig.Scopes
+	}
+
+	httpClient, err := utils.SetupOAuthClient(ctx, config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to setup OAuth client: %w", err)
+	}
+
+	return httpClient, nil
+}
+
+// registerWithSmo sends a message to the SMO to register our identifiers and URL
+func (t *reconcilerTask) registerWithSmo(ctx context.Context) error {
+	// Retrieve the local cluster id value.  It appears to always be identified as "version" in its metadata
+	clusterId, err := utils.GetClusterId(ctx, t.client, "version")
+	if err != nil {
+		return fmt.Errorf("failed to get cluster id: %w", err)
+	}
+
+	httpClient, err := t.setupOAuthClient(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to setup oauth client: %w", err)
+	}
+
+	data := utils.AvailableNotification{
+		GlobalCloudId: t.object.Spec.CloudId,
+		OCloudId:      clusterId,
+		ImsEndpoint:   fmt.Sprintf("https://%s/o2ims-infrastructureInventory/v1", t.object.Spec.IngressHost),
+	}
+
+	body, err := json.Marshal(data)
+	if err != nil {
+		return fmt.Errorf("failed to marshal AvailableNotification: %s", err.Error())
+	}
+
+	smo := t.object.Spec.SmoConfig
+	url := fmt.Sprintf("%s%s", smo.Url, smo.RegistrationEndpoint)
+	result, err := httpClient.Post(url, "application/json", bytes.NewReader(body))
+	if err != nil {
+		return fmt.Errorf("failed to send registration request to '%s': %s", url, err.Error())
+	}
+
+	if result.StatusCode != http.StatusOK {
+		return fmt.Errorf("registration request failed to '%s', HTTP code=%s", url, result.Status)
+	}
+
+	return nil
+}
+
+// setupSmo executes the high-level action set register with the SMO and set up the related conditions accordingly
+func (t *reconcilerTask) setupSmo(ctx context.Context) (err error) {
+	if t.object.Spec.SmoConfig != nil {
```

Comment:
I think the readability of this function would be increased if it were slightly refactored to something like:
```
if t.object.Spec.SmoConfig == nil {
    meta.Remove...
    return nil
}

if utils.IsSmoRegistrationCompleted(t.object) {
    return nil
}

....
return nil
```
Totally up to you if you want to keep it as it is.

Reply:
Fixed